File: app/src/main/java/com/popalay/yocard/data/models/Card.java
package com.popalay.yocard.data.models;

import android.support.annotation.DrawableRes;
import android.support.annotation.IntDef;

import com.github.nitrico.lastadapter.StableId;
import com.popalay.yocard.R;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import io.card.payment.CreditCard;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class Card extends RealmObject implements StableId {

    public static final String ID = "id";
    public static final String HOLDER_NAME = "holderName";

    @PrimaryKey private long id;

    private String number;

    private String redactedNumber;

    private String holderName;

    @CardType private int type;

    @CardColor private int color;

    public Card() {
    }

    public Card(CreditCard creditCard) {
        this.number = creditCard.getFormattedCardNumber();
        this.redactedNumber = creditCard.getRedactedCardNumber();
        switch (creditCard.getCardType()) {
            case MASTERCARD:
                this.type = CARD_TYPE_MASTERCARD;
                break;
            case VISA:
                this.type = CARD_TYPE_VISA;
                break;
            case MAESTRO:
                this.type = CARD_TYPE_MAESTRO;
                break;
        }
    }

    public Card(String number, String redactedNumber, String holderName, @CardType int type) {
        this.number = number;
        this.redactedNumber = redactedNumber;
        this.holderName = holderName;
        this.type = type;
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public String getNumber() {
        return number;
    }

    public void setNumber(String number) {
        this.number = number;
    }

    public String getRedactedNumber() {
        return redactedNumber;
    }

    public void setRedactedNumber(String redactedNumber) {
        this.redactedNumber = redactedNumber;
    }

    public String getHolderName() {
        return holderName;
    }

    public void setHolderName(String holderName) {
        this.holderName = holderName;
    }

    @CardType
    public int getType() {
        return type;
    }

    public void setType(@CardType int type) {
        this.type = type;
    }

    public int getColor() {
        return color;
    }

    public void setColor(int color) {
        this.color = color;
    }

    @DrawableRes
    public int getIconRes() {
        switch (type) {
            case CARD_TYPE_MAESTRO:
                return R.drawable.ic_maestro;
            case CARD_TYPE_MASTERCARD:
                return R.drawable.ic_mastercard;
            case CARD_TYPE_VISA:
                return R.drawable.ic_visa;
            default:
                return 0;
        }
    }

    @DrawableRes
    public int getColorRes() {
        switch (color) {
            case CARD_COLOR_GREEN:
                return R.drawable.bg_card_green;
            case CARD_COLOR_PURPLE:
                return R.drawable.bg_card_purple;
            case CARD_COLOR_RED:
                return R.drawable.bg_card_red;
            default:
            case CARD_COLOR_GREY:
                return R.drawable.bg_card_grey;
        }
    }

    @Override
    public long getStableId() {
        return id;
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({CARD_TYPE_MAESTRO, CARD_TYPE_MASTERCARD, CARD_TYPE_VISA})
    public @interface CardType {}

    public static final int CARD_TYPE_MAESTRO = 0;
    public static final int CARD_TYPE_MASTERCARD = 1;
    public static final int CARD_TYPE_VISA = 2;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({CARD_COLOR_GREY, CARD_COLOR_RED, CARD_COLOR_GREEN, CARD_COLOR_PURPLE})
    public @interface CardColor {}

    public static final int CARD_COLOR_GREY = 0;
    public static final int CARD_COLOR_RED = 1;
    public static final int CARD_COLOR_GREEN = 2;
    public static final int CARD_COLOR_PURPLE = 3;

}

File: data/src/main/kotlin/com/popalay/cardme/data/models/Card.kt
package com.popalay.cardme.data.models

import android.support.annotation.DrawableRes
import android.support.annotation.IntDef
import com.google.gson.annotations.Expose
import com.popalay.cardme.data.R
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey


open class Card(
        @Expose @PrimaryKey var number: String = "",
        @Expose var title: String = "",
        @Expose var redactedNumber: String = "",
        @Expose @field:CardType var cardType: Long = 0L,
        @Expose var generatedBackgroundSeed: Long = System.nanoTime(),
        @Expose var position: Int = 0,
        @Expose var isTrash: Boolean = false,
        @Expose var holder: Holder = Holder()
) : RealmObject(), StableId {

    companion object {
        const val NUMBER = "number"
        const val HOLDER_NAME = "holder.name"
        const val POSITION = "position"
        const val IS_TRASH = "isTrash"

        @Retention(AnnotationRetention.SOURCE)
        @IntDef(CARD_TYPE_MAESTRO, CARD_TYPE_MASTERCARD, CARD_TYPE_VISA, CARD_TYPE_AMEX,
                CARD_TYPE_DINERSCLUB, CARD_TYPE_DISCOVER, CARD_TYPE_JCB)
        annotation class CardType

        const val CARD_TYPE_MAESTRO = 0L
        const val CARD_TYPE_MASTERCARD = 1L
        const val CARD_TYPE_VISA = 2L
        const val CARD_TYPE_AMEX = 3L
        const val CARD_TYPE_DINERSCLUB = 4L
        const val CARD_TYPE_DISCOVER = 5L
        const val CARD_TYPE_JCB = 6L
        const val CARD_TYPE_UNKNOWN = 7L
    }

    @DrawableRes fun getIconRes() = when (cardType) {
        CARD_TYPE_MAESTRO -> R.drawable.ic_maestro
        CARD_TYPE_MASTERCARD -> R.drawable.ic_mastercard
        CARD_TYPE_VISA -> R.drawable.ic_visa
        CARD_TYPE_AMEX -> R.drawable.ic_amex
        CARD_TYPE_DINERSCLUB -> R.drawable.ic_diners_club
        CARD_TYPE_DISCOVER -> R.drawable.ic_discover
        CARD_TYPE_JCB -> R.drawable.ic_jcb
        else -> R.drawable.ic_unknown
    }

    override val stableId: Long
        get() = number.hashCode().toLong()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other?.javaClass != javaClass) return false

        other as Card

        if (number != other.number) return false
        if (title != other.title) return false
        if (redactedNumber != other.redactedNumber) return false
        if (cardType != other.cardType) return false
        if (generatedBackgroundSeed != other.generatedBackgroundSeed) return false
        if (position != other.position) return false
        if (isTrash != other.isTrash) return false

        return true
    }

    override fun hashCode(): Int {
        var result = number.hashCode()
        result = 31 * result + title.hashCode()
        result = 31 * result + redactedNumber.hashCode()
        result = 31 * result + cardType.hashCode()
        result = 31 * result + generatedBackgroundSeed.hashCode()
        result = 31 * result + position
        result = 31 * result + isTrash.hashCode()
        return result
    }

    override fun toString() = "Card(number='$number', title='$title', redactedNumber='$redactedNumber', cardType=$cardType, generatedBackgroundSeed=$generatedBackgroundSeed, position=$position, isTrash=$isTrash)"

}